pkg/kube/kube_handler.go

```diff
@@ -726,6 +727,266 @@ func doValidationRequest(cli httpclient.Client, req *http.Request) (*ValidationR
+		log.Infof("repoURL : %v, repoName : %v\n", repoURL, repoName)
+		tagVersionPtr, err := getOCIAppRepositoryTag(cli, repoURL, repoName)
+		if err != nil {
+			return false, err
+		}
+
+		log.Infof("tagVersion: %v", *tagVersionPtr)
+
+		mediaTypePtr, err := getOCIAppRepositoryMediaType(cli, repoURL, repoName, *tagVersionPtr)
+		if err != nil {
+			return false, err
+		}
+
+		log.Infof("mediaType : %v", *mediaTypePtr)
+
+		wantStr := "vnd.cncf.helm.config"
+
+		result, err = compareMediaType(repoName, *mediaTypePtr, wantStr)
+		if !result {
+			return result, err
+		}
+	}
+	return result, err
+}
+
+type nonOCIRepoRequests struct {
+	reqs []*http.Request
+}
+
+type OCIRepoRequests struct {
+	appRepo *v1alpha1.AppRepository
+}
+
+type HttpValidator interface {
+	IsValid(cli httpclient.Client) (*ValidationResponse, error)
+}
+
+func (nonOCIRepo_reqs nonOCIRepoRequests) IsValid(cli httpclient.Client) (*ValidationResponse, error) {
+	reqs := nonOCIRepo_reqs.reqs
+	var err = error(nil)
+
+	response := &ValidationResponse{}
+
+	for _, req := range reqs {
+		res, err := cli.Do(req)
+		if err != nil {
+			// If the request fail, it's not an internal error
+			return &ValidationResponse{Code: 400, Message: err.Error()}, nil
+		}
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to parse validation response. Got: %v", err)
+		}
+		response = &ValidationResponse{Code: res.StatusCode, Message: string(body)}
+		if response.Code != 200 {
+			return response, nil
+		}
+	}
+
+	return response, err
+}
+
+func (OCIRepo_reqs OCIRepoRequests) IsValid(cli httpclient.Client) (*ValidationResponse, error) {
+
+	var response *ValidationResponse
+	response = &ValidationResponse{Code: 200, Message: ""}
+
+	result, err := ValidateOCIAppRepository(OCIRepo_reqs.appRepo, cli)
+	if !result {
+		response = &ValidationResponse{Code: 400, Message: err.Error()}
+	}
+	return response, err
+}
+
+// getValidators return appropriate HttpValidator interface for OCI and non-OCI Repos
+func getValidators(appRepo *v1alpha1.AppRepository, cli httpclient.Client) (HttpValidator, error) {
+	var validator HttpValidator
+
+	result := []*http.Request{}
+	repoURL := strings.TrimSuffix(strings.TrimSpace(appRepo.Spec.URL), "/")
+
+	switch appRepo.Spec.Type {
+	case "oci":
+		// For the OCI case, we want to validate that all the given repositories are valid
+		if len(appRepo.Spec.OCIRepositories) == 0 {
+			return nil, ErrEmptyOCIRegistry
+		}
+		var OCIRepoReq OCIRepoRequests
+		OCIRepoReq.appRepo = appRepo
+		validator = OCIRepoReq
+		break
+	default:
+		parsedURL, err := url.ParseRequestURI(repoURL)
+		if err != nil {
+			return nil, err
+		}
+		parsedURL.Path = path.Join(parsedURL.Path, "index.yaml")
+		req, err := http.NewRequest("GET", parsedURL.String(), nil)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, req)
+
+		var nonOCIRepoReq nonOCIRepoRequests
+		nonOCIRepoReq.reqs = result
+		validator = nonOCIRepoReq
```

Comment:
Same here:
```suggestion
		return nonOCIRepoReq{
			reqs: requests,
		}
```

I think the compiler will recognise that all possibilities are exhausted, so you don't need the final return below, but check.